File: clients/python/examples/test_watch.py
"""test_watch.py — demonstrates the Watch streaming RPC.

Prerequisites:
    - FileDB server running: `make run` from the repo root.
    - Client installed:      `pip install .` from clients/python.

Run:
    python examples/test_watch.py
"""

import threading
import time

from filedbv2 import FileDB

COLLECTION = "watch_test_py"
INSERT_COUNT = 5


def main() -> None:
    db = FileDB("localhost", 5433, "dev-key")

    # Clean up from any previous run.
    if COLLECTION in db.list_collections():
        db.drop_collection(COLLECTION)
    db.create_collection(COLLECTION)

    print(f'Watching "{COLLECTION}" — will insert {INSERT_COUNT} records...')

    # Insert in a background thread so the main thread can consume the stream.
    def inserter() -> None:
        # A separate client/channel for the writer.
        writer = FileDB("localhost", 5433, "dev-key")
        for i in range(1, INSERT_COUNT + 1):
            time.sleep(0.3)
            rid = writer.insert(
                COLLECTION, {"msg": f"record-{i}", "ts": time.time()}
            )
            print(f"  [insert] id={rid}")
        writer.close()

    threading.Thread(target=inserter, daemon=True).start()

    # Collect exactly INSERT_COUNT watch events, then stop.
    received = 0
    for event in db.watch(COLLECTION):
        print(
            f"  [watch]  op={event['op']} id={event['record']['id']} "
            f"data={event['record']['data']}"
        )
        received += 1
        if received >= INSERT_COUNT:
            break

    db.drop_collection(COLLECTION)
    db.close()
    print("Watch test done.")


if __name__ == "__main__":
    main()

File: clients/js/examples/test_watch.ts
/**
 * test_watch.ts — demonstrates the Watch streaming RPC.
 *
 * Prerequisites:
 *   - ScrivaDB server running: `make run` from the repo root
 *
 * Run:
 *   npx ts-node examples/test_watch.ts
 */

import { ScrivaDB } from '../src';

const COLLECTION = 'watch_test_js';
const INSERT_COUNT = 5;

async function main(): Promise<void> {
  const db = new ScrivaDB('localhost', 5433, 'dev-key');

  // Clean up from any previous run.
  const existing = await db.listCollections();
  if (existing.includes(COLLECTION)) {
    await db.dropCollection(COLLECTION);
  }
  await db.createCollection(COLLECTION);

  console.log(`Watching "${COLLECTION}" — will insert ${INSERT_COUNT} records...`);

  // Start background inserts with a 300ms interval.
  let insertCount = 0;
  const inserter = setInterval(async () => {
    if (insertCount >= INSERT_COUNT) {
      clearInterval(inserter);
      return;
    }
    insertCount++;
    const id = await db.insert(COLLECTION, {
      msg: `record-${insertCount}`,
      ts: new Date().toISOString(),
    });
    console.log(`  [insert] id=${id}`);
  }, 300);

  // Collect exactly INSERT_COUNT watch events, then break.
  let received = 0;
  for await (const event of db.watch(COLLECTION)) {
    console.log(`  [watch]  op=${event.op} id=${event.record.id} data=${JSON.stringify(event.record.data)}`);
    if (++received >= INSERT_COUNT) break;
  }

  clearInterval(inserter);
  await db.dropCollection(COLLECTION);
  db.close();
  console.log('Watch test done.');
}

main().catch((err: unknown) => {
  console.error(err);
  process.exit(1);
});
